src/polybinedges.jl

```diff
@@ -0,0 +1,60 @@
+"""
+    BinEdges <: AbstractVector{Float64}
+
+    This type implements a vector-like data structure to be used for histogram bin edges, it can handle both uniform and non-uniform binnings in a single type to reduce the amount of parametric types. It would switch to O(1) `searchsortedlast` if the binning is uniform.
+
+!!! note
+    Due to the usage of Float64, bin edges shouldn't contain element with absolute value larger than 9007199254740992, which is the `maxintfloat(Float64)`.
+"""
+struct BinEdges <: AbstractVector{Float64}
+    isuniform::Bool
+    nonuniform_edges::Vector{Float64}
+    uniform_edges::StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}
+    inv_step::Float64
+    rfirst::Float64
+    function BinEdges(edges)
+        if isempty(edges)
+            throw(ArgumentError("BinEdges cannot be empty"))
+        end
+        if edges isa AbstractRange
+            new(true, Float64[], edges, inv(step(edges)), first(edges))
+        elseif edges isa AbstractVector
+            if !issorted(edges)
+                throw(ArgumentError("BinEdges must be sorted"))
+            end
+            if !allunique(edges)
+                throw(ArgumentError("BinEdges must be unique"))
+            end
+            if any(x -> abs(x) > maxintfloat(Float64), edges)
+                throw(ArgumentError("BinEdges cannot contain element with absolute value larger than $(maxintfloat(Float64))"))
+            end
+            new(false, edges, 0.0:-1.0, Inf, -Inf)
+        end
+    end
+end
+
+Base.@constprop :aggressive isuniform(b::BinEdges) = b.isuniform
```

Comment:
@Moelf  I am curious. What is the effect of this `@constprop` ? where do you gain performance?

Reply:
If the constant is propagated, then the if-else will be selected at compile time basically (i.e. which searchsorted method)

But yeah I should verify this is actually doing something